### PR TITLE
enhance resource name validation

### PR DESCRIPTION
**Pull request by gronke**
addresses #483 

- be more strict with jail names
- allow certain special characters in jail names `! ^ - _ ( ) [ ] { } < > , .`

### CODE REVIEW

**Comment by gronke** [comment]
### @ Character
@urosgruber the `@` character was not included in the permitted characters, because it is reserved to identify jail snapshots. Allowing the character would complicate the detection of snapshots, which becomes more important when supporting auto-completion.

### Dots in jail names
The dot `.` character has also not been not permitted in jail names, because it is used by FreeBSD jails to express structures of nested jails. The only occasion where this becomes a problem for ioc is when assigning a name parameter to FreeBSD jails (currently `ioc-<JAIL_NAME>`) - without this human-readable identifier /usr/sbin/jls becomes less comfortable to use. This would also break compatibility with older iocage versions. (running jails are identified by this name)

I do understand that it is inviting to name jails after their domain name, but it comes with another disadvantage: the sorting order. Personally I prefer to name jails in reverse FQDN hierarchy (the domain `subdomain.bsd.ci` would turn into `ci-bsd-subdomain`.

The solution implemented in iocage (Python) https://github.com/iocage/iocage/pull/935/files is not acceptable, because it causes name collisions between jails. The two jails `foo.bar` and `foo_bar` would have the same jail identifier, which causes a race condition.

---

It could make sense to sacrifice the `@` character as replacement for dot `.` in the jail identifier. @joachimmathes might suggest change to iocage. This is already a compatibility issue with latest iocgae that would be nice to resolve.

**Comment by gronke** [comment]
Update: dots in jail names are supported now, replacing the character in jail identifiers (Jail param `name`) with `%`, as it appears to be more useful for jail given names. Opinions welcome!

**Comment by igalic** [comment]
> I do understand that it is inviting to name jails after their domain name, but it comes with another disadvantage: the sorting order. Personally I prefer to name jails in reverse FQDN hierarchy (the domain subdomain.bsd.ci would turn into ci-bsd-subdomain.

if you're making such opinionated decisions, it would be good to anchor them in the documentation (and `ioc command help` output)

File: libioc/errors.py [hunk 1]
```diff
@@ -393,7 +393,8 @@ def __init__(
     ) -> None:
         msg = (
             f"Invalid jail name '{name}': "
-            "Names may only contain alphanumeric characters and dash"
+            "Names may only contain alphanumeric characters and may not begin "
+            "with special characters ! ^ - _ ( ) [ ] { } < > , ."
```
Comment (igalic) [inline libioc/errors.py]: it's unclear from this message, whether or not the name is allowed to *contain* the these characters.

File: tests/test_Jail.py [hunk 2]
```diff
@@ -339,3 +340,55 @@ def test_hooks_can_access_ioc_env_variables(
...
+        assert libioc.helpers.is_valid_name("ioc\t23") is False
+        assert libioc.helpers.is_valid_name("ioc-") is False
+        assert libioc.helpers.is_valid_name("!ioc") is False
+        assert libioc.helpers.is_valid_name("@gronke") is False
+        assert libioc.helpers.is_valid_name("stefan@gronke") is False
+        assert libioc.helpers.is_valid_name("23-ioc-") is False
+        assert libioc.helpers.is_valid_name("white space") is False
+        assert libioc.helpers.is_valid_name("ioc\\jail") is False
+        assert libioc.helpers.is_valid_name("jail{ioc}") is False
+        assert libioc.helpers.is_valid_name("we❤jails") is False
```
Comment (igalic) [inline tests/test_Jail.py]: sad times 🙀

File: libioc/errors.py (hunk 1 above)
Comment (gronke) [reply libioc/errors.py]: The error message is clarified now. Thanks for spotting it!

**Comment by gronke** [comment]
Jail names may now contain dots (e.g. `jail.example.com`). The implementation deviates from python-iocage, which replaces dots in jail name parameters (used as identifier of managed jails) with dash (`-`). Because dash is allowed in regular jail names, the running state of two jails `my.jail` and `my-jail` would be confused. This implementation replaces dots in FreeBSD jail names with `*` - a character that is not allowed in ioc Jail names.

**Pull request merged**